Document ECP_RESTARTABLE and make it compatible with USE_PSA

mpg opened the pull request:
## Description

This PR does two things:

1. Fix a documentation gap: there was no place except the ChangeLog that mentioned the effect of `MBEDTLS_ECP_RESTARTABLE` on TLS.
2. Remove a long-standing limitation by making it possible to build with both `ECP_RESTARTABLE` and `USE_PSA`. The result for TLS is not fully satisfying, but limitations are documented and tested, and this is still an improvement over the existing (mostly for non-TLS use of restartable).



## Gatekeeper checklist

- [x] **changelog** provided
- [x] **backport** #6753 - just the first part (documentation fix & test improvement)
- [x] **tests** provided

AndrzejKurek commented on include/mbedtls/mbedtls_config.h:
```diff
@@ -690,11 +690,33 @@
+ *   - ECDH operations from the key exchange;
+ *   - verification of the server's key exchange signature;
+ *   - verification of the server's certificate chain;
+ *   - generation of our signature if client authentication is used, with an
```
"our" in this context sounds a bit strange. Client signature?

AndrzejKurek commented on library/pk_wrap.c:
```diff
@@ -1162,8 +1162,13 @@ static int ecdsa_sign_wrap( void *ctx_arg, mbedtls_md_type_t md_alg,
     size_t key_len;
     unsigned char buf[MBEDTLS_PK_ECP_PRV_DER_MAX_BYTES];
     unsigned char *p;
+#if defined(MBEDTLS_ECDSA_DETERMINISTIC)
```
Commit c0b47d5e87290ea83ff7a5e00c177b64f8300736 has a couple of misspellings:
- `Use deterministic in PSA when we do in legacy` <- deterministic ECDSA?
- `The failure was cause by` <- caused
It would be great to know what test function(s?) you're talking about to look at more details.

AndrzejKurek replied:
This is still present, but is a minor.

AndrzejKurek commented on docs/use-psa-crypto.md:
```diff
@@ -86,28 +83,34 @@ is enabled, no change required on the application side.
+- Verification of RSA-PSS signatures with a salt length that is different from
   the hash length.
+- Restartable operations when `MBEDTLS_ECP_RESTARTABLE` is also enabled (see
+  the documentation of that option).
```
Maybe also a pointer to `psa-limitations.md`?

AndrzejKurek replied:
Ah, I can see that it's redundant. No need to include it too.

AndrzejKurek commented on tests/suites/test_suite_ecp.function:
```diff
@@ -237,7 +237,7 @@ exit:
 /* END_CASE */
 
-/* BEGIN_CASE depends_on:MBEDTLS_ECP_RESTARTABLE */
+/* BEGIN_CASE depends_on:MBEDTLS_ECP_RESTARTABLE:MBEDTLS_ECP_SHORT_WEIERSTRASS_ENABLED */
```
Finally something useful found by `depends.py`! :D

AndrzejKurek reviewed (approved):
Looks good given its limitations (that seem to be well documented).
Minor nitpicks to fix if you consider them reasonable.

The only part that IMO might need a changelog entry is the part about using deterministic ECDSA in `ecdsa_sign_wrap`, as it changes previous behaviour under the hood.

gilles-peskine-arm reviewed (changes requested):
Looks mostly good to me, just a few remarks about wording.

mpg commented on include/mbedtls/mbedtls_config.h:
```diff
@@ -690,11 +690,33 @@
+ *        MBEDTLS_ECP_ALT, MBEDTLS_ECDH_XXX_ALT, MBEDTLS_ECDSA_XXX_ALT,
+ *        and MBEDTLS_USE_PSA_CRYPTO.
+ *
+ * Uncomment this macro to enable restartable ECC computations.
```
Yes, I'm not sure I like it, but most options have something like that as the last line, so I kept if for consistency.

mpg commented:
@gilles-peskine-arm @AndrzejKurek Thanks for your review. I've addressed your feedback by adding 3 commits at the end, then edited the history to include the missing information in the commit message of what's now 79ae7eb. The commit you previously reviewed was e1653de.

AndrzejKurek commented on ChangeLog.d/pk-sign-restartable.txt:
```diff
@@ -0,0 +1,5 @@
+Changes
+   * When MBEDTLS_USE_PSA_CRYPTO and MBEDTLS_ECDSA_DETERMINISTIC are both
+     defined, mbedtls_pk_sign() now use deterministic ECDSA for ECDSA
```
Minor
```suggestion
     defined, mbedtls_pk_sign() now uses deterministic ECDSA for ECDSA
```

AndrzejKurek reviewed (approved):
Only three typos found.

gilles-peskine-arm reviewed (approved):
LGTM
